server/sonar-server/src/main/java/org/sonar/server/computation/measure/MeasureDtoToMeasure.java

```diff
@@ -0,0 +1,111 @@
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.computation.measure;
+
+import com.google.common.base.Optional;
+import java.util.Objects;
+import javax.annotation.Nullable;
+import org.sonar.core.measure.db.MeasureDto;
+import org.sonar.server.computation.metric.Metric;
+
+import static org.sonar.server.computation.measure.Measure.AlertStatus.toAlertStatus;
+
+public class MeasureDtoToMeasure {
+
+  public Optional<Measure> toMeasure(@Nullable MeasureDto measureDto, Metric metric) {
+    Objects.requireNonNull(metric);
+    if (measureDto == null) {
+      return Optional.absent();
+    }
+
+    Double value = measureDto.getValue();
+    String data = measureDto.getData();
+    switch (metric.getMetricType().getValueType()) {
+      case INT:
+        return toIntegerMeasure(measureDto, value, data);
+      case LONG:
+        return toLongMeasure(measureDto, value, data);
+      case DOUBLE:
+        return toDoubleMeasure(measureDto, value, data);
+      case BOOLEAN:
+        return toBooleanMeasure(measureDto, value, data);
+      case STRING:
+        return toStringMeasure(measureDto, data);
+      case NO_VALUE:
+        return toNoValueMeasure(measureDto);
+      default:
+        throw new IllegalArgumentException("Unsupported Measure.ValueType " + metric.getMetricType().getValueType());
+    }
+  }
+
+  private static Optional<Measure> toIntegerMeasure(MeasureDto measureDto, @Nullable Double value, String data) {
+    if (value == null) {
+      return toMeasure(MeasureImpl.createNoValue(), measureDto);
+    }
+    return toMeasure(MeasureImpl.create(value.intValue(), data), measureDto);
+  }
+
+  private static Optional<Measure> toLongMeasure(MeasureDto measureDto, @Nullable Double value, String data) {
+    if (value == null) {
+      return toMeasure(MeasureImpl.createNoValue(), measureDto);
+    }
+    return toMeasure(MeasureImpl.create(value.longValue(), data), measureDto);
+  }
+
+  private static Optional<Measure> toDoubleMeasure(MeasureDto measureDto, @Nullable Double value, String data) {
+    if (value == null) {
+      return toMeasure(MeasureImpl.createNoValue(), measureDto);
+    }
+    return toMeasure(MeasureImpl.create(value.doubleValue(), data), measureDto);
+  }
+
+  private static Optional<Measure> toBooleanMeasure(MeasureDto measureDto, @Nullable Double value, String data) {
+    if (value == null) {
+      return toMeasure(MeasureImpl.createNoValue(), measureDto);
+    }
+    return toMeasure(MeasureImpl.create(value == 1.0d, data), measureDto);
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Equality tests should not be made with floating point values. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1244)